### PR TITLE
DNN-8968: user target user's preferred locale to get notification content.

### PR DESCRIPTION
**Pull request by zyhfish**

### CODE REVIEW

File: DNN Platform/Tests/DotNetNuke.Tests.Integration/Tests/FriendsTests.cs [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+using DotNetNuke.Security.Profile;
+using DotNetNuke.Services.Cache;
+using DotNetNuke.Services.Localization;
+using NUnit.Framework;
+
+namespace DotNetNuke.Tests.Integration.Tests
+{
+    public class FriendsTests : IntegrationTestBase
+    {
+        #region Fields
```
Comment (tpluscode) [inline DNN Platform/Tests/DotNetNuke.Tests.Integration/Tests/FriendsTests.cs]: Let's stop using regions please

File: DNN Platform/Library/Entities/Users/Social/FriendsControllerImpl.cs [hunk 2]
```diff
@@ -143,10 +143,11 @@ public void DeleteFriend(UserInfo initiatingUser, UserInfo targetUser)
         private static void AddFriendRequestNotification(UserInfo initiatingUser, UserInfo targetUser, UserRelationship userRelationship)
         {
             var notificationType = NotificationsController.Instance.GetNotificationType(FriendRequest);
-            var subject = string.Format(Localization.GetString("AddFriendRequestSubject", Localization.GlobalResourceFile),
+            var language = GetUserPreferredLocale(targetUser)?.Name;
```
[the remark this answers is not in the document]
Comment (zyhfish) [reply DNN Platform/Library/Entities/Users/Social/FriendsControllerImpl.cs]: The Localization.GetString already have fallback language logic, don't need add it here.

**Comment by zyhfish** [comment]
This test case need much pre-conditions, i think integration test is more efficient than unit test, whats your opinion? thx.

**Comment by tpluscode** [comment]
Of course, I see how it requires all kinds of set up for languages and users.

At the same time I see the call to `EventManager.Instance.OnFriendshipRequested`. Would it make it possible to move the implementation of `AddFriendRequestNotification` out of the controller and send the notification from a handler of that event?

In such case a unit test would only have to set up `NotificationsController.Instance`, `LocaleController.Instance` and `LocalizationProvider.Instance`.

Please tell me if it's possible

**Comment by zyhfish** [comment]
it maybe possible, but im not sure whether we need do that huge work for this only functionality test.

**Comment by tpluscode** [comment]
That's a weird thing to say. In general unit tests should be those easier to write because there are less part involved, less set up. And the benefit is that they are less prone to breaking because of that. And would also run faster. In general unit tests should be the most numerous IMO.

In this particular case creating a handler and testing that in isolation does not look like huge work and the tests will be much simpler. What I don't see is how it should then be attached to that event.

**Pull request merged**